compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -828,18 +829,32 @@ checkDefTypeSyn DefTypeSyn{synParams,synType} = do
   where
     base = checkType synType KStar
 
+
 -- | Check that an interface definition is well defined.
 checkIface :: MonadGamma m => Module -> DefInterface -> m ()
-checkIface m DefInterface{..} = do
-  checkUnique (EDuplicateInterfaceChoiceName intName) $ NM.names intFixedChoices
-  checkUnique (EDuplicateInterfaceMethodName intName) $ NM.names intMethods
-  forM_ intRequires (inWorld . lookupInterface) -- verify that required interface exists
-  forM_ intMethods checkIfaceMethod
+checkIface m iface = do
+  let tcon = Qualified PRSelf (moduleName m) (intName iface)
+
+  -- check requires
+  when (tcon `S.member` intRequires iface) $
+    throwWithContext (ECircularInterfaceRequires (intName iface) Nothing)
+  forM_ (intRequires iface) $ \requiredIfaceId -> do
+    requiredIface <- inWorld (lookupInterface requiredIfaceId)
+    let missing = intRequires requiredIface `S.difference` intRequires iface
+    when (tcon `S.member` intRequires requiredIface) $
+      throwWithContext (ECircularInterfaceRequires (intName iface) (Just requiredIfaceId))
+    whenJust (listToMaybe (S.toList missing)) $ \missingIfaceId ->
+      throwWithContext (ENotClosedInterfaceRequires (intName iface) requiredIfaceId missingIfaceId)
```

Comment:
Perhaps it would be better to show all missing required interfaces in one go? I think this should do the trick:

```suggestion
    whenJust (nonEmpty (S.toList missing)) $ \missingIfaceIds ->
      throwWithContext (ENotClosedInterfaceRequires (intName iface) requiredIfaceId missingIfaceIds)
```
(together with the corresponding change for `ENotClosedInterfaceRequires`)

Reply:
Not sure it is worthwhile to improve the error right now. We are not sure this error will stay in the final version.

Reply:
good idea, changed the error to take a list!

Reply:
nice, thank you!